datastore/longhorn.go

```diff
@@ -3429,3 +3455,167 @@ func (s *DataStore) ListOrphansByNodeRO(name string) ([]*longhorn.Orphan, error)
 func (s *DataStore) DeleteOrphan(orphanName string) error {
 	return s.lhClient.LonghornV1beta2().Orphans(s.namespace).Delete(context.TODO(), orphanName, metav1.DeleteOptions{})
 }
+
+type Disk struct {
+	spec   *longhorn.DiskSpec
+	status *longhorn.DiskStatus
+}
+
+func (s *DataStore) IsStorageSpaceSufficientForSnapshotPurge(volumeName string) (bool, error) {
+	replicas, err := s.ListReplicasByVolume(volumeName)
+	if err != nil {
+		return false, err
+	}
+
+	disks, err := s.listDisksContainingReplicas(replicas)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to list disks containing replicas for volume %v", volumeName)
+	}
+
+	requiredPurgeReservedSpace, err := s.getRequiredPurgeReservedSpace(volumeName)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get required reserved space for volume %v", volumeName)
+	}
+
+	// Walk through each disk containing the replica of the volume and
+	// check if the disk space enough for snapshot purge.
+	for diskID, disk := range disks {
+		if ok, err := s.isDiskSpaceSufficientForSnapshotPurge(diskID, disk, requiredPurgeReservedSpace); !ok {
+			return false, errors.Wrapf(err, "failed to check disk %v space for snapshout purge for volume %v", diskID, volumeName)
+		}
+	}
+
+	return true, nil
+}
+
+func (s *DataStore) listDisksContainingReplicas(replicas map[string]*longhorn.Replica) (map[string]*Disk, error) {
+	disks := map[string]*Disk{}
+
+	for _, replica := range replicas {
+		disk, err := s.getDisk(replica.Spec.NodeID, replica.Spec.DiskID)
+		if err != nil {
+			return nil, err
+		}
+		disks[replica.Spec.DiskID] = disk
+	}
+	return disks, nil
+}
+
+// Walk through the engines of the given volume and find the largest snapshot size which
+// is regarded as the required reserved space for the volume to be purged.
+func (s *DataStore) getRequiredPurgeReservedSpace(volumeName string) (int64, error) {
+	engines, err := s.ListEnginesByVolume(volumeName)
+	if err != nil {
+		return 0, err
+	}
+
+	largestSnapshotSize := int64(0)
+	for _, engine := range engines {
+		size, err := getLargestSnapshotSize(engine)
```

Comment:
Should we consider different reserved size for different types of purge? For the current calculation, it seems for coalescing. 

```golang
		case replica.OpCoalesce:
		case replica.OpRemove:
		case replica.OpReplace:
		case replica.OpPrune:
```

Reply:
Can we get this op info in the longhorn manager? It seems that these are mainly used inside the engine.

Reply:
probably not..

Reply:
Yes. And I prefer not to make this check too complicated.

Reply:
It's clear now 🙂